libsolidity/analysis/TypeChecker.cpp

```diff
@@ -338,35 +340,62 @@ bool TypeChecker::visit(FunctionDefinition const& _function)
 		if (_function.isOrdinary() && !_function.isPartOfExternalInterface())
 			m_errorReporter.typeError(5587_error, _function.location(), "Internal functions cannot be payable.");
 	}
-	auto checkArgumentAndReturnParameter = [&](VariableDeclaration const& var) {
-		if (type(var)->containsNestedMapping())
-			if (var.referenceLocation() == VariableDeclaration::Location::Storage)
+
+	vector<VariableDeclaration const*> internalParametersInConstructor;
+
+	auto checkArgumentAndReturnParameter = [&](VariableDeclaration const& _var) {
+		if (type(_var)->containsNestedMapping())
+			if (_var.referenceLocation() == VariableDeclaration::Location::Storage)
 				solAssert(
-					_function.libraryFunction() || !_function.isPublic(),
+					_function.libraryFunction() || _function.isConstructor() || !_function.isPublic(),
 					"Mapping types for parameters or return variables "
 					"can only be used in internal or library functions."
 				);
-		if (_function.isPublic())
+		bool functionIsExternallyVisible =
+			(!_function.isConstructor() && _function.isPublic()) ||
+			(_function.isConstructor() && !m_currentContract->abstract());
+		if (
+			_function.isConstructor() &&
+			_var.referenceLocation() == VariableDeclaration::Location::Storage &&
+			!m_currentContract->abstract()
+		)
+			m_errorReporter.typeError(
+				3644_error,
+				_var.location(),
+				"This parameter has a type that can only be used internally. "
+				"You can make the contract abstract to avoid this problem."
+			);
+		else if (functionIsExternallyVisible)
 		{
-			auto iType = type(var)->interfaceType(_function.libraryFunction());
+			auto iType = type(_var)->interfaceType(_function.libraryFunction());
 
 			if (!iType)
 			{
-				solAssert(!iType.message().empty(), "Expected detailed error message!");
-				m_errorReporter.typeError(4103_error, var.location(), iType.message());
+				string message = iType.message();
+				solAssert(!message.empty(), "Expected detailed error message!");
+				if (_function.isConstructor())
+					message += " You can make the contract abstract to avoid this problem.";
+				m_errorReporter.typeError(4103_error, _var.location(), message);
+			}
+			else if (
+				!experimentalFeatureActive(ExperimentalFeature::ABIEncoderV2) &&
+				!typeSupportedByOldABIEncoder(*type(_var), _function.libraryFunction())
+			)
+			{
+				string message =
+					"This type is only supported in ABIEncoderV2. "
+					"Use \"pragma experimental ABIEncoderV2;\" to enable the feature.";
+				if (_function.isConstructor())
+					message +=
+						" Alternatively, make the contract abstract and supply the "
+						"constructor arguments from a derived contract.";
+				m_errorReporter.typeError(
+					4957_error,
```

Comment:
Isn't this error the same as the one above, just from a different reason?

Reply:
Good point! I'll try to see what happens.

Reply:
Nope, `interfaceType` turns it into a memory reference - so we actually need this additional check.

Reply:
Ok!